### DISCUSSION TITLE
Why such a new C++ version?

### DISCUSSION THREAD

**Discussion by swagtoy** [WerWolv/ImHex#1205] [General]
Look, I understand this is your project and your codebase... but why are we using such a modern C++ version?

Generally, I tend to hold off a few years before switching to a new C++ version. I specifically look for things like Debian and what not to support these. 

The problem though, hurts things like OpenBSD which is still using Clang 13 (this was release 2 years ago, sure, but even that supports some C++20...), but we are already jumping the gun and using C++23... Even worse, C++23 isn't even "released" yet! Why are we jumping the gun and slurping up what compilers give out???

I feel like hopping on new versions seems a bit harsh because it limits who can use the software, and I generally feel like some of these changes are rather small and more of a bragging point to say "Look at me, I'm using new C++ features!". Most C++ projects still hang around C++17, and often older.

This is wonderful software, but it's annoying having to dual boot into Debian to use it, and I assume that I'll only get to use it for so long before my compiler becomes outdated on Debian too.

I wish you would rely more on libraries like fmt or something.

**Comment by WerWolv** [comment]
Hey

There's really no reason other than "I like playing around with new features". I generally hate the mentality of holding back with using new features just because some distros refuse to update their compiler. I know this is a big thing with C and C++,  but in most modern language people just use the latest versions available and aren't at the grace of whatever distro they're using.

Normally I recommend people to just use the AppImage or Flatpak version. I have no idea about OpenBSD though so not sure if you can use those. Maybe you can build a version that works for you using the Dockerfile we provide?

**Reply by swagtoy** [reply]
> I generally hate the mentality of holding back with using new features just because some distros refuse to update their compiler.

The issue isn't just features, but it's also stability. People don't just upgrade because it seems better. Some distros rely on older compilers because new ones introduced something new or broke portability in some strange way temporarily.

OpenBSD tends to follow the latter, it's a stability preference. You're also just beta testing new C++ features and you're never 100% sure those work everywhere. It's just asking for trouble in the long run. Docker is also only a Linux thing. 

I understand playing around with new features, but do consider that there are other people who want to actually use the software. You limit your audience by playing with new features and risk possible bugs. The end goal of software is to make it run, I don't think making it your sandbox is the greatest idea (but then again, it's your codebase, you're free to do what you want).

Of course, you could also keep specific version maybe that you maintain for "stability" reasons and occasionally apply patches too? That would allow you to keep flowing with new C++ versions while providing a "good enough" version that is a bit older.

---

The ONLY way for me to port this software to OpenBSD atm is to modify the cmake file to not depend on some repositories and modify some code to not use such modern features. It's tedious, but it's all that is available

**Reply by WerWolv** [reply]
I honestly had very different experiences. Distros refusing to update their compiler and rather shipping with versions where some features are just completely broken. It's super frustrating to just not use stuff that's available and works perfectly fine just because some distros just rather stick with older versions where it doesn't work yet.

I used the latest available versions since day one of ImHex and they never caused any issues in the past 3 years. The biggest issue always was e.g old Ubuntu versions just not updating their stuff.
That's the reason why we bundle e.g Capstone, libfmt or nlohmann-json. Those libraries also were mostly available everywhere but some were 5 years behind and had major bugs in them that just weren't fixed.

In any case, I'm happy to make changes to ImHex to make building more straight forward for FreeBSD and OpenBSD. I am not however moving back to outdated compiler versions.

**Reply by swagtoy** [reply]
> I am not however moving back to outdated compiler versions.

At the very least, I am asking to move back to C++20. C++20 is actually standardized and supported well by old compilers. C++23, on the other hand, isn't officially released yet you are already claiming that this goes up to C++23. 

Im not trying to say "old compilers", i'm trying to say "old C++ version". C++20 isn't that old.

OpenBSD supports C++20, not C++23

**Comment by iTrooz** [comment]
Hey, just wanted to say that I would be glad to to make ImHex work on OpenBSD (probably with the AppImage), if you currently have no way to run it (I have no idea about the current support status, but you seem to imply that it doesn't wofk)

Feel free to ping me if that's the case

**Reply by swagtoy** [reply]
OpenBSD doesn't support AppImages. AppImages are a Linux-only thing.

FreeBSD appears to be the only *BSD that supports ImHex, but this was very recent and it seems to patch quite a few things to get it working

**Reply by iTrooz** [reply]
Hey, could you check out https://github.com/WerWolv/ImHex/issues/1206 when you get time please ?
